src/rust/catmem/queue.rs

```diff
@@ -5,68 +5,292 @@
 // Imports
 //======================================================================================================================
 
-use super::pipe::Pipe;
+use super::{
+    ring::Ring,
+    QMode,
+};
 use crate::{
-    collections::shared_ring::SharedRingBuffer,
     runtime::{
         fail::Fail,
+        limits,
+        memory::DemiBuffer,
         queue::IoQueue,
         QType,
     },
     scheduler::{
         TaskHandle,
+        Yielder,
         YielderHandle,
     },
 };
-use ::std::collections::HashMap;
+use ::std::{
+    cell::RefCell,
+    collections::HashMap,
+    rc::Rc,
+};
+
+//======================================================================================================================
+// Constants
+//======================================================================================================================
+
+/// Maximum number of retries for pushing a EoF signal.
+const MAX_RETRIES_PUSH_EOF: u32 = 16;
 
 //======================================================================================================================
 // Structures
 //======================================================================================================================
 
-/// Per-queue metadata: reference to the shared pipe.
+/// This structure contains code for manipulating a single, Catmem-specific Demikernel queue. Queue state is kept in
+/// the [ring] structure, while [pending_ops] holds the map of TaskHandles and YielderHandle for currently active async
+/// functions.
+#[derive(Clone)]
 pub struct CatmemQueue {
-    pipe: Pipe,
-    pending_ops: HashMap<TaskHandle, YielderHandle>,
+    ring: Rc<RefCell<Ring>>,
+    pending_ops: Rc<RefCell<HashMap<TaskHandle, YielderHandle>>>,
 }
 
 //======================================================================================================================
 // Associated Functions
 //======================================================================================================================
 
 impl CatmemQueue {
-    pub fn new(ring: SharedRingBuffer<u16>) -> Self {
-        Self {
-            pipe: Pipe::new(ring),
-            pending_ops: HashMap::<TaskHandle, YielderHandle>::new(),
+    /// This function creates a new CatmemQueue and a new shared ring buffer and connects to it to either the consumer
+    /// or producer end indicated by [mode].
+    pub fn create(name: &str, mode: QMode) -> Result<Self, Fail> {
+        match mode {
+            QMode::Push => Ok(Self {
+                ring: Rc::new(RefCell::<Ring>::new(Ring::create_push_ring(name)?)),
+                pending_ops: Rc::new(RefCell::<HashMap<TaskHandle, YielderHandle>>::new(HashMap::<
```

Comment:
The creation of pending_ops is repeated. Possibly, it can be moved outside of the match statements?

Reply:
I'm sure the compiler will deduplicate it but do you think it would be easier to read?

Reply:
Its better for readability and maintenance - fewer lines of code to read as well as change, if required. Plus, the decision should only apply to how the `ring` is created, and not how `pending_ops` is created. [Not a blocker]

Reply:
Sure, but I don't agree with the argument presented here... On the one hand we are shrinking one line of code, on the other we increasing the scope of `pending_ops`  decoupling resorce creation from resource binding. This can lead to misuses in the future that may incurr in side-effects...

Still, I am applying this change